app/com/linkedin/drelephant/tez/fetchers/TezFetcher.java

```diff
@@ -0,0 +1,481 @@
+        if( finishTime < firstAttemptFinishTime) {
+          firstAttempt = taskAttempt;
+          firstAttemptFinishTime = finishTime;
+        }
+      }
+      return firstAttempt;
+    }
+
+
+
+    private void getTaskData(String dagId, List<TezTaskData> taskList, boolean isMapTask)
+        throws IOException, AuthenticationException {
+
+      if (taskList.size() > MAX_SAMPLE_SIZE) {
+        logger.info(dagId + " needs sampling.");
+        Collections.shuffle(taskList);
+      }
+
+      int sampleSize = Math.min(taskList.size(), MAX_SAMPLE_SIZE);
+      for (int i=0; i<sampleSize; i++) {
+        TezTaskData data = taskList.get(i);
+        URL taskCounterURL = getTaskURL(data.getTaskId());
+        TezCounterData taskCounter = getTaskCounter(taskCounterURL);
+
+        URL taskAttemptURL = getTaskAttemptURL(dagId, data.getTaskId(), data.getAttemptId());
+        long[] taskExecTime = getTaskExecTime(taskAttemptURL, isMapTask);
+
+        data.setCounter(taskCounter);
+        data.setTime(taskExecTime);
+      }
+
+    }
+
+    private TezCounterData getTaskCounter(URL url) throws IOException, AuthenticationException {
+      JsonNode rootNode = ThreadContextMR2.readJsonNode(url);
+      JsonNode groups = rootNode.path("otherinfo").path("counters").path("counterGroups");
+      TezCounterData holder = new TezCounterData();
+
+      //Fetch task level metrics
+      for (JsonNode group : groups) {
+        for (JsonNode counter : group.path("counters")) {
+          String name = counter.get("counterName").getTextValue();
+          String groupName = group.get("counterGroupName").getTextValue();
+          Long value = counter.get("counterValue").getLongValue();
+          holder.set(groupName, name, value);
+        }
+      }
+
+      return holder;
+    }
+
+    private long[] getTaskExecTime(URL url, boolean isMapTask) throws IOException, AuthenticationException {
+      JsonNode rootNode = ThreadContextMR2.readJsonNode(url);
+      JsonNode groups = rootNode.path("otherinfo").path("counters").path("counterGroups");
+
+      long startTime = rootNode.path("otherinfo").get("startTime").getLongValue();
+      long finishTime = rootNode.path("otherinfo").get("endTime").getLongValue();
+
+      long shuffleTime = 0;
+      long mergeTime = 0;
+
+      for (JsonNode group : groups) {
+        for (JsonNode counter : group.path("counters")) {
+          String name = counter.get("counterName").getTextValue();
+          if (!isMapTask && name.equals("MERGE_PHASE_TIME")) {
+            mergeTime = counter.get("counterValue").getLongValue();
+          }
+          else if (!isMapTask && name.equals("SHUFFLE_PHASE_TIME")){
+            shuffleTime = counter.get("counterValue").getLongValue();
+          }
+
+        }
+      }
+
+      long[] time = new long[] { finishTime - startTime, shuffleTime, mergeTime, startTime, finishTime };
+
+      return time;
+    }
+  }
+}
+
+final class ThreadContextMR2 {
```

Comment:
ThreadContextMR2 is also used by MapReduceFetcherHadoop2. Refactor code here so there is only one instance of  ThreadContextMR2.